### DISCUSSION TITLE
Debugging RadioLib on STM32F103C8 with sx1262 - MCU Hang-ups

### DISCUSSION THREAD

**Discussion by SamLue** [jgromes/RadioLib#1120] [General]
I use the RadioLib in the current version 6.6.0 on a STM32F103C8 with a RA-01SH (sx1262). I soldered the hardware together myself. First I did tests with TTN/LoRaWan and joined successfully and sent payload cyclically. At first I did not notice any problems. But then I rewrote my code and sent normal LoRa data.
setup

```
void setup() {
...
int state = radio.begin(868.0,125.0,7,5,RADIOLIB_SX126X_SYNC_WORD_PRIVATE,10,8,0,false);
...

void loop() {
....
int transmissionState = radio.startTransmit(uplinkPayload, sizeof(uplinkPayload));
...
}

```
The first LoRa message is also sent. And often the second one too. But actually either after the first or after the second message, everything hangs up. The MCU stops working.

OK, so I wanted to analyze the reason and activate debugging.

But as soon as I comment out the line.

`#define RADIOLIB_DEBUG_BASIC  `

Everything hangs at the radio.begin.
```

    Serial1.println(F("Initialise the radio"));
    int state = radio.begin(868.0,125.0,7,5,RADIOLIB_SX126X_SYNC_WORD_PRIVATE,10,8,0,false);
    debug(state != RADIOLIB_ERR_NONE, F("Initialise radio failed"), state, true);

    Serial1.println(F("Ready!\n"));
```

I can still see "Initialise the radio" but neither “Ready” nor an error.
What could the reason be for this behavior?

**Comment by jgromes** [comment]
I can think of two things that can be wrong, though I have not seen your entire code:

1. You seem to be usiung `Serial1` as the debug output - RadioLib assumes `Serial` by default. It can be changed by defining `RADIOLIB_DEBUG_PORT  Serial1` in oyur build system or in BuildOptUser.h
2. I'm guessing that `debug` function si from LoRaWAN examples `config.h`. Since it halts on error, it is possible the STM does not have enough time to print out the error message. So you might want to either set the last argument of `debug` to false, or add some waiting delay into the infinite while loop inside debug. This function will likely get removed from the examples in future updatres anyway.

**Reply by SamLue** [reply]
Hi Jan,

first of all, thank you for your great work on this library and also for your helpful answer here.

Good point about Serial1 and thanks for the variable RADIOLIB_DEBUG_PORT. When I define this to Serial1, the behavior is exactly as expected according to when I don't enable debug.

However, your second point does not apply. Even if I do not use the debug function, the behavior remains identical.

But I have now approached the issue from the other side and have completely discarded my code and adapted and tested the SX126x_Transmit_Interrupt.ino example with this hardware. And lo and behold, nothing hangs up and the packets are sent. So the hardware is fine and not the reason. So I must be making a very stupid mistake in my code, which I now have to look for and will find.

**Reply by SamLue** [reply]
I have found the misstake. It was something with transmittedFlag and the radio.setPacketSentAction(setFlag);
At the moment I don't know where exactly but the time between two startTransmit is > 1 minute and big enough.

Thanks again for all your work!

**Comment by ShreyaVadla** [comment]
Hi SamLue.

How did you use this radio library with the STM and SX1262, since it is Arduino-based? Did you port it to STM? I want to use this radio library for LoRaWAN with a Nucleo F746 and SX1262 but don't know the procedure. I apologize if I'm mistaken, as I am very new to LoRaWAN. Please correct me if I'm wrong.

**Reply by jgromes** [reply]
@ShreyaVadla 

> STM and SX1262, since it is Arduino-based?

I beg to differ, running this library does not require Arduino at all - see the NonArduino examples. In any case, there is the excellent [stm32duino Arduino core](https://github.com/stm32duino/Arduino_Core_STM32) project, that allows you to use the Arduino API on STM32 boards.

**Reply by SamLue** [reply]
... and this is the way I'm using RadioLib. As Development Environment VS Code with PlatformIO. And in the Project I'm using the Arduino Framework with STSTM32 as platform.


```
[env:nameit]
platform = ststm32
framework = arduino
...

```
